### PR TITLE
Prepare for frozen string literals

### PR DESCRIPTION
**Pull request by HarlemSquirrel**
- Add the magic comment to all files
- Update Overcommit::HookContext::PreCommit to not use encode!
- Set Rubocop target Ruby version to 2.3

### CODE REVIEW

**Comment by sds** [comment]
Can you provide additional justification for why this is worthwhile, @HarlemSquirrel? 

Ruby 2.2 may be EOL but enabling frozen string literals by default is not without its risks. If existing code uses string mutation it will now fail.

**Comment by HarlemSquirrel** [comment]
Freezing mutable strings reduces memory usage. This will be the default in Ruby 3. 

https://github.com/rails/rails/pull/20946

**Comment by sds** [comment]
Hey @HarlemSquirrel, can you get TravisCI tests passing?

File: lib/overcommit/hook_context/pre_commit.rb [hunk 1]
```diff
@@ -19,7 +21,8 @@ def amendment?
       # containing unicode characters, we replace the offending characters,
       # since the pattern we're looking for will consist of ASCII characters
       unless cmd.valid_encoding?
-        cmd.encode!('UTF-16be', invalid: :replace, replace: '?').encode!('UTF-8')
+        cmd = Overcommit::Utils.parent_command.encode('UTF-16be', invalid: :replace, replace: '?')
+                               .encode('UTF-8')
```
[the remark this answers is not in the document]
Comment (HarlemSquirrel) [reply lib/overcommit/hook_context/pre_commit.rb]: Yes. The original code mutated the string while this change sets `cmd` to a new string allowing us to freeze all string literals in this file.

**Pull request merged**